### ISSUE TITLE
Change signing keys without downtime

### ISSUE THREAD

**Issue by sastels** [cds-snc/notification-planning-core#50]
## Description

As a Notify admin, I need to be able to rotate our signing keys without downtime or manual writes to the database.

WHY are we building?
- currently changing our signing keys `SECRET_KEY` and `DANGEROUS_SALT` requires manually resigning several database columns as well as the notifications in redis, and approximately 30-60 minutes of downtime.
- we want to move to a system where signing keys can be rotated without manual database manipulation nor downtime.

WHAT are we building?
As discussed in the ADR,
- change our signing salts to be string constants describing the signing context
- change the `SECRET_KEY` to be a list of keys
- create a recurring task to resign the affected tables

VALUE created by our solution
- rotating the signing keys will involve PRs changing `SECRET_KEY`.

## Acceptance Criteria
- [ ] salts changed
- [ ] `SECRET_KEY` is a list
- [ ] scheduled tasks are running

## QA Steps
- [ ] Have the resigning tasks run. Shouldn't make any changes. Verify that Notify still works.
- [ ] Rotate `SECRET_KEY` array as discussed in the ADR. Verify that Notify still works.
- [ ] Have the resigning tasks run.  Verify that Notify still works.
- [ ] Rotate `SECRET_KEY` array again. At this point the original value of `SECRET_KEY` should have rotated out entirely.  Verify that Notify still works.

**Comment by sastels** [comment]
PR to stop using DANGEROUS_SALT in token generation
https://github.com/cds-snc/notification-utils/pull/204

**Comment by sastels** [comment]
still need to switch admin and api to the new utils

**Comment by sastels** [comment]
Switch admin and api to new utils:
https://github.com/cds-snc/notification-admin/pull/1547
https://github.com/cds-snc/notification-api/pull/1863

I haven't tested these yet though.

**Comment by sastels** [comment]
To test in staging:
- change your email address
- send a team invitation

**Comment by sastels** [comment]
code approve, needs to be merged and tested in staging

**Comment by sastels** [comment]
merged into staging., ready for QA

**Comment by sastels** [comment]
Andrew QAd in staging

**Comment by sastels** [comment]
So the app isn't using the DANGEROUS_SALT to create tokens anymore. In 2 days there won't be any valid tokens signed with the DANGEROUS_SALT, so we can stop passing it into utils, and then we can change utils to not want a salt.

At this point the DANGEROUS_SALT will only be used to salt the user passwords before hashing.

**Comment by sastels** [comment]
2 days has passed, we can merge the admin / api PRs into staging and prod as desired.

**Comment by CalvinRodo** [comment]
Is there a way to rotate the salt without expiring passwords? Maybe something like 

```
   Validate password with old salt
   if valid resign with new salt
     if not valid validate with new salt 
   if valid let them login
   if not valid do invalid password stuff
```

**Comment by sastels** [comment]
yep, 100% we could keep the old salt around and re-hash with the new salt (then after a bit, mark any that haven't logged in as expired). Just more code to write and time to spend on it when there's other stuff to do. We've tried to keep the scope creep down to, essentially:
- figure out what all our secrets are (and document)
- rotate without downtime

But totally worth considering as a next improvement. I thought I had added this to the "[Next Steps](https://docs.google.com/document/d/1RXlbgIlYzVY5DMWix81vIZluTp-FBLMz8HCYi-v7I7s/edit#heading=h.4p2n7e1nrep9)" but hadn't yet so I just did.

**Comment by jimleroyer** [comment]
Steve said there are a few items left to work on. We will move back to In Progress.

**Comment by ben851** [comment]
* waiting for a PR to be reviewed @jimleroyer

**Comment by sastels** [comment]
A couple more PRs using the new utils (ie where generate_token() doesn't have a salt argument)
https://github.com/cds-snc/notification-admin/pull/1573
https://github.com/cds-snc/notification-api/pull/1890 <--- needs a review

**Comment by ben851** [comment]
* Almost done, Steve to merge in PRs - we were waiting for the releases today to complete.

**Comment by sastels** [comment]
Still one PR to merge (already approved).
